agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java

```diff
@@ -61,26 +61,30 @@ private void detectVagrant(@NotNull final BuildAgentConfiguration config) {
       }
 
       private void detectDocker(@NotNull final BuildAgentConfiguration config) {
-        if (!SystemInfo.isLinux) {
-          LOG.debug("Docker is only available under Linux");
-          return;
-        }
 
-        final String output = executeCommandWithShell("docker", "docker --version");
+        final String output = executeCommandWithShell("docker", "docker version");
         if (output == null) return;
 
-        String ver = output.toLowerCase().trim();
-        ver = ver.replaceAll("\\s*docker\\s+version\\s+", "");
-        ver = ver.replaceAll(",?\\s+build\\s+", "-");
+        final String dockerVersion = executeCommandWithShell("docker", "docker version --format {{.Server.Version}}");
+        final String dockerHostOS = executeCommandWithShell("docker", "docker version --format {{.Server.Os}}");
 
-        if (StringUtil.isEmptyOrSpaces(ver)) {
-          LOG.warn("Failed to parse docker version: " + output);
+        if (StringUtil.isEmptyOrSpaces(dockerVersion)) {
+          LOG.warn("Failed to parse docker version" + output);
+          return;
+        } else if (StringUtil.isEmptyOrSpaces(dockerHostOS)) {
+          LOG.warn("Failed to parse docker host os" + output);
           return;
         }
 
-        config.addConfigurationParameter(VMConstants.DOCKER_PROPERTY, ver);
+        config.addConfigurationParameter(VMConstants.DOCKER_PROPERTY, cleanOutput(dockerVersion));
+        config.addConfigurationParameter(VMConstants.DOCKER_HOST_OS_PROPERTY, cleanOutput(dockerHostOS));
       }
 
+      private String cleanOutput(String rawOutput) {
+        return rawOutput
+                .trim()
+                .replace(System.getProperty("line.separator"), "");
```

Comment:
I would recommend first replace regex  `[\s\r\n]+` with single a whitespace, next to trim. Currently, for multiline output, it's possible to contact lines without spaces at all.